microcontrollers/_index.md: well supported boards

bxparks opened the pull request:
Add a list of "well supported" microcontrollers (SAMD21, SAMD51, nRF52840, RP2040), and a short list of popular, but not well-supported, microcontrollers (AVR, ESP8266, ESP32). I did not add any recommendations of specific brands (e.g. Adafruit, Sparkfun, Seeed, Wemos, Raspberry, Arduino, etc.) because I think that list would be subjective, and probably needs to be different based on the user's place of residence.

bxparks committed:
microcontrollers/_index.md: well supported boards

Add a list of "well supported" microcontrollers (SAMD21, SAMD51, nRF52840,
RP2040), and a short list of popular, but not well-supported, microcontrollers
(AVR, ESP8266, ESP32). I did not add any recommendations of specific brands
(e.g. Adafruit, Sparkfun, Seeed, Wemos, Raspberry, Arduino, etc.) because I
think that list would be subjective, and probably needs to be different based
on the user's place of residence.

aykevl commented on content/docs/reference/microcontrollers/_index.md:
```diff
@@ -10,6 +10,37 @@ TinyGo lets you run Go directly on microcontrollers.
+  ARM Cortex-M0+ processor
+* [SAMD51](https://www.microchip.com/en-us/product/ATSAMD51N19A) based on the
+  ARM Cortex-M4 processor
+* [nRF52840](https://infocenter.nordicsemi.com/topic/ps_nrf52840/keyfeatures_html5.html)
+  based on the Arm Cortex-M4F processor
```
Most nRF5x boards are well supported:
* nrf52832 also has good support (but is less commonly used by Adafruit and the like)
* nrf52833 is used in the microbit v2, so it probably also works fine
* nrf51822 is older, has less memory, but I expect it also works fine
* there may be other nrf52xxx chips that may not be as well supported, but that need only small tweaks to work well

bxparks replied:
Added "other nRF" boards with references to nrf52833,  nrf52832, nrf51822. I'm not familiar with any of these, and they seem to be rare under `tinygo/targets`:

* grep -i nrf51822 * | wc: 1
* grep -i nrf52832 * | wc: 1
* grep -i nrf52833 * | wc: 8
* grep -i nrf52840 * | wc: 39

aykevl commented on content/docs/reference/microcontrollers/_index.md:
```diff
@@ -10,6 +10,37 @@ TinyGo lets you run Go directly on microcontrollers.
+  ARM Cortex-M4 processor
+* [nRF52840](https://infocenter.nordicsemi.com/topic/ps_nrf52840/keyfeatures_html5.html)
+  based on the Arm Cortex-M4F processor
+* [RP2040](https://en.wikipedia.org/wiki/RP2040) with dual ARM Cortex-M0+
+  processors
```
It's correct that it's dual core, but saying it here may lead people to expect that TinyGo is also able to use both cores (which it isn't). So I'd recommend leaving that out.

bxparks replied:
I think the RP2040 is famous for being a cheap dual core, so people will want to know that. But it's a bit of important information that TinyGo supports only a single core, so I added that as a parenthetical remark.

aykevl commented on content/docs/reference/microcontrollers/_index.md:
```diff
@@ -10,6 +10,37 @@ TinyGo lets you run Go directly on microcontrollers.
+* [RP2040](https://en.wikipedia.org/wiki/RP2040) with dual ARM Cortex-M0+
+  processors
+
+The introductory Arduino boards based on the 8-bit AVR processors work
+relatively well under TinyGo. But they have limited amounts of flash and static
+memory so they support only small applications:
```
True. Perhaps giving a few specific examples might help? For example: you can't use `fmt` on these chips and goroutines don't work so well due to limited RAM.

bxparks replied:
> True. Perhaps giving a few specific examples might help? For example: you can't use `fmt` on these chips and goroutines don't work so well due to limited RAM.

Yup, added

aykevl commented on content/docs/reference/microcontrollers/_index.md:
```diff
@@ -10,6 +10,37 @@ TinyGo lets you run Go directly on microcontrollers.
+As of early 2023, boards using the following microcontrollers are
+well-supported:
+
+* [SAMD21](https://www.microchip.com/en-us/product/ATSAMD21G18) based on the
+  ARM Cortex-M0+ processor
+* [SAMD51](https://www.microchip.com/en-us/product/ATSAMD51N19A) based on the
+  ARM Cortex-M4 processor
```
Perhaps it is useful to give their branding names. The SAMD21 chips are called "M0" by Adafruit and "Zero" by Arduino. The SAMD51 board is called "M4" by Adafruit (Arduino doesn't seem to be selling any SAMD51 boards).

bxparks replied:
> Perhaps it is useful to give their branding names. The SAMD21 chips are called "M0" by Adafruit and "Zero" by Arduino. The SAMD51 board is called "M4" by Adafruit (Arduino doesn't seem to be selling any SAMD51 boards).

Done

aykevl reviewed (commented):
This is great! A few suggestions below.

bgould commented:
I think an important detail for newcomers might that some boards support loading code over USB vs needing a separate programmer.  Similarly, some boards/chips support serial-over-USB without any extra hardware and some will require a separate UART-USB converter.  I assume that most newcomers might prefer for things to "just work" out of the box over a USB connection.

bxparks commented:
>

bxparks commented:
Ugh, accidentally hit the wrong item in the drop down... reopening...

bxparks commented:
> I think an important detail for newcomers might that some boards support loading code over USB vs needing a separate programmer. Similarly, some boards/chips support serial-over-USB without any extra hardware and some will require a separate UART-USB converter. I assume that most newcomers might prefer for things to "just work" out of the box over a USB connection.

That is useful information, but I feel like that is answering a slightly different question. I'm trying to answer: "Which boards are well-supported by TinyGo?" But your information is more along the line of: "What boards should I buy or use?" There seems to be many more factors involved in that answer: cost, number of GPIO ports, I2C support, wifi/bluetooth support, flash memory, static memory, CPU speed, USB-to-serial adapter or USB-CDC, power consumption, physical size of dev board.

I feel like we should create a separate page for that.

bxparks committed:
microcontrollers/_index.md:

Add references to other nRF microcontrollers which apparently work well with
TinyGo. Add comment that TinyGo uses only a single core even on dual-core
processors. Give a couple of examples of TinyGo/Go features which may not work
well under AVR processors with limited resources.

bxparks commented:
Ping.. any other comments?
I think I addressed all the prior feedback, so I think this is ready to be merged.
Just want to clear off my pending PRs...

aykevl commented:
Yes, looks good to me.
